app/src/main/java/app/revanced/integrations/patches/AlternativeThumbnailsPatch.java

```diff
@@ -13,29 +16,183 @@
+        // (as much as 4x slower has been observed, despite the alt webp image being a smaller file).
+
+        StringBuilder builder = new StringBuilder(decodedUrl.sanitizedUrl.length() + 2);
+        builder.append(decodedUrl.urlPrefix);
+        builder.append(decodedUrl.videoId).append('/');
+        builder.append(qualityToUse.getAltImageNameToUse());
+        builder.append('.').append(decodedUrl.imageExtension);
+
+        String sanitizedReplacement = builder.toString();
+        if (!VerifiedQualities.verifyAltThumbnailExist(decodedUrl.videoId, qualityToUse, sanitizedReplacement)) {
+            return decodedUrl.sanitizedUrl;
+        }
+
+        return builder.toString();
+    }
+
+    /**
+     * Build the alternative thumbnail url using DeArrow thumbnail cache.
+     *
+     * @param videoId ID of the video to get a thumbnail of.
+     * @param fallbackUrl URL to fall back to in case.
+     * @return The alternative thumbnail url, without tracking parameters.
+     */
+    @NonNull
+    private static String buildDeArrowThumbnailURL(String videoId, String fallbackUrl) {
+        // Use fallback if parsing API URI failed.
+        if (dearrowApiUri == null) {
+            return fallbackUrl;
+        }
+
+        // Build thumbnail request url.
+        // See https://github.com/ajayyy/DeArrowThumbnailCache/blob/29eb4359ebdf823626c79d944a901492d760bbbc/app.py#L29.
+        return dearrowApiUri
+                .buildUpon()
+                .appendQueryParameter("videoID", videoId)
+                .appendQueryParameter("officialTime", "true")
+                .appendQueryParameter("redirectUrl", fallbackUrl)
+                .build()
+                .toString();
+    }
+
+    /**
+     * Injection point.  Called off the main thread and by multiple threads at the same time.
+     *
+     * @param originalUrl Image url for all url images loaded, including video thumbnails.
+     */
+    public static String overrideImageURL(String originalUrl) {
+        try {
+            final var thumbnailMode = AlternativeThumbnailMode.getCurrent();
+            if (thumbnailMode == AlternativeThumbnailMode.ORIGINAL) {
+                return originalUrl;
+            }
+
+            final var decodedUrl = DecodedThumbnailUrl.decodeImageUrl(originalUrl);
+            if (decodedUrl == null) {
+                return originalUrl; // Not a thumbnail.
+            }
+
+            // Keep any tracking parameters out of the logs, and log only the base URL.
+            LogHelper.printDebug(() -> "Original url: " + decodedUrl.sanitizedUrl);
+
+            final StringBuilder thumbnailUrlBuilder = new StringBuilder();
+            if (thumbnailMode.isDeArrow()) {
+                // Get fallback URL.
+                final String fallbackUrl = thumbnailMode == AlternativeThumbnailMode.DEARROW_OR_VIDEO_STILLS
+                        ? buildYoutubeVideoStillURL(decodedUrl)
+                        : decodedUrl.sanitizedUrl;
```

Comment:
How do you assume that the recommendations are being affected? I don't think the thumbnails are being used to track app usage, as it would be a pretty inflexible and primitive approach.